### DISCUSSION TITLE
is it possible to configure the tool-chain in meson.build?

### DISCUSSION THREAD

**Discussion by stsp** [mesonbuild/meson#12545] [Q&A]
Hi.

I've found the very old thread about that
subject: #1752 but it ended inconclusively.

Today we have multiple tool-chains,
each with its own set of bugs and features.
My project represents a very complex system,
that can only be built using clang++ (not g++)
as a compiler, GNU ld (and not lld) as a linker,
and a customly patched version of nasm
(and not, say, yasm).

I realize that the intention of meson is to
allow the user to select the preferable tool-chain
when he builds the code. But today's tool-chains
are so complex and the new features are added
so rapidly to them, that the assumption that
"every tool-chain suits", is basically no longer valid.

So my question is: does meson allow me to
configure the needed tool-chain for my project
(by selecting the components from multiple
tool-chains and adding custom tools), without
asking the user to please set CC/CXX/LD/NASM
variables on his own?

**Comment by xclaesse** [comment]
You can define your toolchain in a native file: https://mesonbuild.com/Machine-files.html. But that still requires the user to run `meson setup --native-file my_toolchain.ini`.

**Reply by stsp** [reply]
> But that still requires the user to run

And will, if I understand correctly, change
his global setup, so he may eventually get
failures from other projects?

**Reply by xclaesse** [reply]
Main project and all subprojects are built with the same toolchain indeed.

**Reply by stsp** [reply]
I meant, does this setup command
affect all projects of that user (i.e.
a system-wide preferences), or is it
local to the project which he builds
now?

**Reply by eli-schwartz** [reply]
Using a native file happens per project that you run "setup" on, it is not a system wide setting.

**Reply by stsp** [reply]
Good to know, thanks.
I probably can then use a small
"configure" script, which will just
configure meson rightly.

Have found that meson doesn't
support flex/bison natively, although
its already suggested in this bugzilla.
Ok, I probably need to start trying
things out. :)

**Reply by eli-schwartz** [reply]
There's an example of using flex/bison in https://github.com/mesonbuild/meson/blob/master/test%20cases/frameworks/8%20flex/meson.build

The basic idea is to define a generator for it.

**Reply by stsp** [reply]
Yes, I see, thanks.
Its just that I am evaluating a
couple of build systems at once,
and for example with xmake I
can do:
```
target("parser")
    add_rules("lex", "yacc")
    add_files("parser.y", "parser.l")
```
So the native support is always
better. But I can use the custom
rule too, as its not the biggest
part of the challenge anyway.

**Comment by stsp** [comment]
https://github.com/mesonbuild/meson/pull/10938
https://github.com/mesonbuild/meson/pull/9379#issuecomment-1071904847
Seems these are related.
But not implemented as yet?
